File: commands.py
import pandas as pd
from Bio.PDB import *
import numpy as np
import json
import seaborn as sns
import pickle


# monomer
## read af output 

features=pd.read_pickle("full_outputs/alphafold/features.pkl")
features.keys()
features["template_sequence"]
features["template_domain_names"]

## model rankings
with open('full_outputs/alphafold/ranking_debug.json') as json_file:
    data = json.load(json_file)
    
data

best_model=pd.read_pickle("full_outputs/alphafold/result_model_2.pkl")

best_model.keys()
best_model["plddt"]
best_model_plddt=pd.DataFrame({"residue":range(len(best_model["plddt"])), "plddt":best_model["plddt"]})
sns.scatterplot(data=best_model_plddt, x="residue", y="plddt")


## load best model with ptm
best_ptm=pd.read_pickle("full_outputs/alphafold/result_model_3_ptm.pkl")
sns.heatmap(best_ptm["predicted_aligned_error"])


## rosettafold outputs
rosetta_params=np.load("full_outputs/rosettafold/t000_.3track.npz", mmap_mode="r")

for key in rosetta_params.keys():
    print(key)

# multimer outputs

with open("multimer/term/result_model_1_multimer.pkl", "rb") as pick:
    params=pickle.load(pick)

params.keys()

features=pd.read_pickle("multimer/term/features.pkl")

multi_plddt=pd.DataFrame({"residue":range(len(params["plddt"])), "plddt":params["plddt"], "chain":features["entity_id"]})
multi_plddt["protein"]="eRF1"
multi_plddt.loc[multi_plddt["chain"]==2, "protein"]="eRF3"
sns.scatterplot(data=multi_plddt, x="residue", y="plddt", hue="protein")

sns.heatmap(data=params["predicted_aligned_error"])

# calculate contacts

parser = PDBParser()
erf1_erf3=parser.get_structure(id="erf1_erf3", file="multimer/term/ranked_0.pdb")
chains=Selection.unfold_entities(erf1_erf3, "C")
chains

erf1=chains[0]
erf3=chains[1]

backbone_atoms=[]
for residue in erf1:
    backbone_atoms.append(residue["CA"])
    
ns=NeighborSearch(backbone_atoms)

erf3_contacts=[]

for residue in erf3:
    backbone_atom=residue["CA"]
    close_atoms=ns.search(backbone_atom.coord, 10)
    if len(close_atoms) > 0:
        erf3_residues=[atom.get_parent() for atom in close_atoms]
        erf3_contacts=erf3_contacts+erf3_residues
    

erf3_contacts=list(set(erf3_contacts))
erf3_contacts


erf1_abce1=parser.get_structure(id="erf1_erf3", file="multimer/recyc/ranked_0.pdb")


chains=Selection.unfold_entities(erf1_abce1, "C")

erf1=chains[0]
abce1=chains[1]

backbone_atoms=[]
for residue in erf1:
    backbone_atoms.append(residue["CA"])
    
ns=NeighborSearch(backbone_atoms)


abce1_contacts=[]

for residue in abce1:
    backbone_atom=residue["CA"]
    close_atoms=ns.search(backbone_atom.coord, 10)
    if len(close_atoms) > 0:
        abce1_residues=[atom.get_parent() for atom in close_atoms]
        abce1_contacts=abce1_contacts+abce1_residues
    

abce1_contacts=list(set(abce1_contacts))
abce1_contacts

same_residues=[residue for residue in abce1_contacts if residue in erf3_contacts]
same_residues.sort()
same_residues


# separate pdb into domans
parser = PDBParser()
wt=parser.get_structure(id="wt", file="monomer/multi_domain/model_1.crderr.pdb")

domains=pd.read_csv("monomer/multi_domain/DomainSpreadsheet-Homo_sapiens_Transcript_Domains_ENST00000545968.csv")
gene3d=domains[domains["Domain source"]=="Gene3D"]

starts=gene3d["Start"].tolist()
ends=gene3d["End"].tolist()

len(starts)==len(ends)
gene3d=domains[domains["Domain source"]=="Gene3D"]

class GetDomain(Select):
    def __init__(self, start, end):
        self.start=start
        self.end=end
        
    def accept_residue(self, residue):
        if residue.id[1] >= self.start and residue.id[1]<=self.end:
            return 1
        else:
            return 0

for i in range(len(starts)):
    filename="domain_"+str(i)+".pdb"
    io=PDBIO()
    io.set_structure(wt)
    io.save(filename, GetDomain(start=starts[i],end=ends[i]))

File: commands.R
library(pheatmap)
library(reticulate)
library(ggplot2)
library(dplyr)
library(jsonlite)
library(bio3d)

# load alphafold model

pd<-import("pandas")
features<-pd$read_pickle("full_outputs/alphafold/features.pkl")

str(features, max.level = 1)

models<-fromJSON("full_outputs/alphafold/ranking_debug.json")
models

best_model=pd$read_pickle("full_outputs/alphafold/result_model_2.pkl")
names(best_model)

best_model_plddt<-data.frame(residue=1:length(best_model$plddt), 
  plddt=best_model$plddt)

ggplot(best_model_plddt, aes(x=residue, y=plddt))+geom_point()

best_model_ptm<-pd$read_pickle("full_outputs/alphafold/result_model_3_ptm.pkl")
pheatmap(best_model_ptm$predicted_aligned_error, cluster_cols=F, cluster_rows=F)

# load rosetta model
np<-import("numpy")

rosetta_params<-np$load("full_outputs/rosettafold/t000_.3track.npz", mmap_mode="r")
rosetta_params$files
rosetta_params$f["dist"]

# multimer outputs

params=pd$read_pickle("multimer/term/result_model_1_multimer.pkl")
features=pd.read_pickle("multimer/term/features.pkl")

multi_plddt<-data.frame(residue=1:length(params$plddt), 
  plddt=params$plddt, chain=features$entity_id)

multi_plddt$protein<-ifelse(multi_plddt$chain==1, "eRF1", "eRF3")

ggplot(multi_plddt, aes(x=residue, y=plddt, color=protein))+geom_point()

pheatmap(params$predicted_aligned_error, cluster_cols=F, cluster_rows=F)

# load pdb plot beta factors
pdb<-read.pdb("multimer/term/ranked_0.pdb")

ggplot(unique(pdb$atom[, c("resno", "b", "chain")]), aes(x=resno, y=b, color=chain))+geom_point()+
  facet_grid(.~chain)


# calculate contacts

inds <- atom.select(pdb, "calpha")
contacts <- cmap(pdb, inds=atom.select(pdb, "calpha"), dcut=10, binary=F, collapse=F, mask.lower=F)

contacts_long<-reshape2::melt(contacts)

#select erf1 and erf3 residues
contacts_filt<-contacts_long %>%
  filter(Var1 %in% multi_plddt$residue[multi_plddt$protein=="eRF1"]) %>%
  filter(Var2 %in% multi_plddt$residue[multi_plddt$protein=="eRF3"]) %>%
  filter(value==1)


erf1_contacts<-unique(contacts_filt$Var1)
erf3_contacts<-unique(contacts_filt$Var2)

# separate into domains

pdb<-read.pdb("monomer/multi_domain/model_1.crderr.pdb")
domains<-read.table("monomer/multi_domain/DomainSpreadsheet-Homo_sapiens_Transcript_Domains_ENST00000545968.csv", 
  header=T, sep=",") %>%
  filter(Domain.source=="Gene3D")



for(i in 1:nrow(domains)){
  inds<-atom.selecT(pdb, resno=domains$Start[i]:domains$End[i])
  domain_atoms<-trim.pdb(pdb, inds)
  write.pdb(domain_atoms, file=paste0("domain_", i, ".pdb"))
}












